drivers/i2c/i2c_mchp_mss.c

```diff
@@ -0,0 +1,384 @@
+
+	return 0;
+}
+
+static int mss_i2c_write(const struct device *dev, uint8_t serial_addr, uint8_t *tx_buffer,
+			 uint32_t tx_num_write)
+{
+	struct mss_i2c_data *const data = dev->data;
+	const struct mss_i2c_config *cfg = dev->config;
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	atomic_t target_status = data->target_status;
+
+	if (data->transaction == NO_TRANSACTION) {
+		data->transaction = CONTROLLER_WRITE_TRANSACTION;
+	}
+
+	data->pending_transaction = CONTROLLER_WRITE_TRANSACTION;
+	data->target_addr = serial_addr << TARGET_ADDR_SHIFT;
+	data->dir = I2C_MSG_WRITE;
+	data->controller_tx_buffer = tx_buffer;
+	data->controller_tx_size = tx_num_write;
+	data->controller_tx_idx = 0u;
+	atomic_set(&data->controller_status, MSS_I2C_IN_PROGRESS);
+
+	if (target_status == MSS_I2C_IN_PROGRESS) {
+		data->is_transaction_pending = CONTROLLER_WRITE_TRANSACTION;
+	} else {
+		sys_write8((ctrl | CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+	}
+
+	if (data->bus_status == MSS_I2C_HOLD_BUS) {
+		sys_write8((ctrl & ~CTRL_SI), cfg->i2c_base_addr + CORE_I2C_CTRL);
+	}
+
+	return 0;
+}
+
+
+static int mss_i2c_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			    uint16_t addr)
+{
+	for (int i = 0; i < num_msgs; i++) {
+		struct i2c_msg *current = &msgs[i];
+
+		if ((current->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+			mss_i2c_read(dev, addr, current->buf, current->len);
+			mss_wait_complete(dev);
+		} else {
+			mss_i2c_write(dev, addr, current->buf, current->len);
+			mss_wait_complete(dev);
+		}
+	}
+
+	return 0;
+}
+
+static const struct i2c_driver_api mss_i2c_driver_api = {
+	.configure = mss_i2c_configure,
+	.transfer = mss_i2c_transfer,
+};
+
+static void mss_i2c_reset(const struct device *dev)
+{
+	const struct mss_i2c_config *cfg = dev->config;
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	sys_write8((ctrl & ~CTRL_ENS1), cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	sys_write8((ctrl | CTRL_ENS1), cfg->i2c_base_addr + CORE_I2C_CTRL);
+}
+
+
+static void mss_i2c_irq_handler(const struct device *dev)
+{
+	struct mss_i2c_data *const data = dev->data;
+	const struct mss_i2c_config *cfg = dev->config;
+
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	uint8_t status = sys_read8(cfg->i2c_base_addr + CORE_I2C_STATUS);
+
+	uint8_t hold_bus = 0;
+
+	switch (status) {
+	case STATUS_M_START_SENT:
+	case STATUS_M_REPEATED_START_SENT:
+		sys_write8((ctrl & ~CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+		sys_write8(data->target_addr | data->dir, cfg->i2c_base_addr + CORE_I2C_DATA);
+
+		data->controller_tx_idx = 0;
+		data->controller_rx_idx = 0;
+
+		data->is_transaction_pending = false;
+		data->transaction = data->pending_transaction;
+		break;
+	case STATUS_M_ARB_LOST:
+		sys_write8((ctrl | CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+		printk("lost arbitration: %x\n", status);
```

Comment:
Isn't better to use `LOG_ERR` or so?

Reply:
Hey @edersondisouza, thanks for taking the time to review. I don't think LOG_ERR is the right choice here. From reading the Logging API it says that LOG_ERR "It's meant to report severe errors, such as those from which it's not possible to recover." 
When that state occurs the IP actually releases the bus and goes into target mode.  

Would you think LOG_DBG or LOG_WRN might suite?

Cheers,
Conor

Reply:
@edersondisouza and done! thanks for pointing that one out.